File: src/main/java/de/pflugradts/passbird/domain/service/password/MovePasswordService.java
package de.pflugradts.passbird.domain.service.password;

import com.google.inject.Inject;
import de.pflugradts.passbird.domain.model.PasswordEntryNotFound;
import de.pflugradts.passbird.domain.model.namespace.NamespaceSlot;
import de.pflugradts.passbird.domain.model.password.KeyAlreadyExistsException;
import de.pflugradts.passbird.domain.model.transfer.Bytes;
import de.pflugradts.passbird.domain.service.eventhandling.EventRegistry;
import de.pflugradts.passbird.domain.service.password.encryption.CryptoProvider;
import de.pflugradts.passbird.domain.service.password.storage.PasswordEntryRepository;

public class MovePasswordService implements CommonPasswordServiceCapabilities {

    @Inject
    private CryptoProvider cryptoProvider;
    @Inject
    private PasswordEntryRepository passwordEntryRepository;
    @Inject
    private EventRegistry eventRegistry;

    public void movePassword(final Bytes keyBytes, final NamespaceSlot targetNamespace) {
        if (entryExists(cryptoProvider, passwordEntryRepository, keyBytes, targetNamespace)) {
            throw new KeyAlreadyExistsException(keyBytes);
        } else {
            var encryptedKeyBytes = encrypted(cryptoProvider, keyBytes);
            find(passwordEntryRepository, encryptedKeyBytes)
                .ifPresentOrElse(
                passwordEntry -> passwordEntry.updateNamespace(targetNamespace),
                () -> eventRegistry.register(new PasswordEntryNotFound(encryptedKeyBytes)));
            processEventsAndSync(eventRegistry, passwordEntryRepository);
        }
    }

}

File: src/main/kotlin/de/pflugradts/passbird/domain/service/password/MovePasswordService.kt
package de.pflugradts.passbird.domain.service.password

import de.pflugradts.kotlinextensions.TryResult
import de.pflugradts.passbird.domain.model.egg.EggIdAlreadyExistsException
import de.pflugradts.passbird.domain.model.event.EggNotFound
import de.pflugradts.passbird.domain.model.shell.Shell
import de.pflugradts.passbird.domain.model.slot.Slot
import de.pflugradts.passbird.domain.service.eventhandling.EventRegistry
import de.pflugradts.passbird.domain.service.password.encryption.CryptoProvider
import de.pflugradts.passbird.domain.service.password.tree.EggRepository
import jakarta.inject.Inject

class MovePasswordService @Inject constructor(
    cryptoProvider: CryptoProvider,
    eggRepository: EggRepository,
    eventRegistry: EventRegistry,
) : CommonPasswordServiceCapabilities(cryptoProvider, eggRepository, eventRegistry) {
    fun movePassword(eggIdShell: Shell, targetSlot: Slot): TryResult<Unit> {
        if (eggExists(eggIdShell, targetSlot)) {
            throw EggIdAlreadyExistsException(eggIdShell)
        } else {
            findWithoutUpdatingMemory(eggIdShell).ifPresentOrElse(
                {
                    eggRepository.discardFavorites(it.associatedNest(), it.viewEggId())
                    it.moveToNestAt(targetSlot)
                },
                { eventRegistry.register(EggNotFound(eggIdShell)) },
            )
            return processEventsAndSync()
        }
    }
}
